test/skate_web/controllers/detour_missed_stops_controller_test.exs

```diff
@@ -0,0 +1,72 @@
+defmodule SkateWeb.DetourMissedStopsControllerTest do
+  use SkateWeb.ConnCase
+
+  import Test.Support.Helpers
+  import Skate.Factory
+
+  alias Skate.Detours.MissedStops
+
+  describe "missed_stops/2" do
+    @tag :authenticated
+    test "returns missed stops", %{conn: conn} do
+      route_pattern = build(:gtfs_route_pattern)
+      shape_with_stops = build(:shape_with_stops)
+      missed_stop = Enum.at(shape_with_stops.stops, 0)
+
+      connection_start = shape_with_stops.points |> Enum.at(0) |> Util.Location.as_location!()
+      connection_end = shape_with_stops.points |> Enum.at(1) |> Util.Location.as_location!()
+
+      reassign_env(:skate_web, :route_pattern_fn, fn _ -> route_pattern end)
+      reassign_env(:skate_web, :shape_with_stops_fn, fn _ -> shape_with_stops end)
+
+      reassign_env(:skate_web, :missed_stops_fn, fn %MissedStops{
+                                                      connection_start: ^connection_start,
```

Comment:
How come you mocked the algorithm, rather than just providing some values for `route_pattern` and `shape_with_stops`, etc, that have simple enough expected results?

I'm partly asking because mocking out the internal function makes this test kind of brittle, and makes the whole PR a bit tougher to review. If we update `missed_stops_fn` to return data in a different way, then this test will still pass, and won't tell us anything about how well this controller works in context.

Reply:
See if this was TypeScript I'd just make sure that it was set up such that the mocked function has to have the same type signature as the original, meaning that any changes to the interface without updating the mocked function would trigger a type error...

In any event I'm not sure we could elegantly do that in Elixir / Dialyzer except for _maybe_ via behaviours. My concern with leaving the test to call `missed_stops/1` directly is that it's a very complicated function and mocking out input data and expected output data would involve a lot of geospatial reasoning that seems pretty disconnected from the controller itself. I think overall the current approach is the lesser evil given that the return type of the function (list of stops) seems like it will be fairly stable.

Reply:
It's not exactly blocking my review of this (it does make reviewing it a bit more complicated, because I have to cross-reference this mock with the behavior of `missed_stops`), but at some point, would you object if I tried refactoring this test to use "realistic" test data instead? Getting reasonable fake data that abstracts away a lot of the geospatial reasoning is a good use case for factories, and we're starting to build the beginnings of a good factory system.

Basically, I'm not as convinced that this is the lesser of two evils, but I don't think it's worth blocking this PR on, and that's kind of a philosophical point anyway.

Reply:
Yeah I'd certainly check out the result if you want to try that!